### ISSUE TITLE
Enhancement: [naming-convention] add support for `#private` members

### ISSUE THREAD

**Issue by hlysine** [typescript-eslint/typescript-eslint#6258]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.4&sourceType=module&code=MYGwhgzhAEAqCmEAu0DeBYAUNaB6X0SAFovNAJYAm8AdkuQGbnwBOMERA9gK4iXQ1OKALZgkwItAacWopDADu5YoRLQARAAduAIxDlg66MM6VGzFlhwBiTS3IA3MfABizPtAC80AIwBuK2hbeyckeABZeGJTAAoASjQAXyxEoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6JgQwFtmBzWsgHsmAN0QtKw9GADa4bDkTRog6JAA08hVgXZIAM1VcO+aTMgBlfFTL4ACh2RkO8AMKPEkALqbde5IhItqrSkGTwjsh2KqTQRBpaupBcggAmlPqUSqgY5sSwAEaMZN6JEAC+vro6fgZGJmaQyNaUtu5cge4B3lW1AUH4IRhhEchRMUrxZWCV0zVJhtDGprmWLbYOTi5dnj7TOP2IwWrD4ZEAsoj4ABZpCX44KemZ2Y35Ra2lfrN+8wp1Swaq2aNnw7U6Hh6+ygh2OoTOY0uNzu03KiS88jR5SAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
class Test {
  // these identifiers should not match formats with the "public" modifier
  #privateField = 1;
  #privateMethod() {}
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "format": ["StrictPascalCase"],
        "selector": "classProperty",
        "modifiers": ["public"]
      },
      {
        "format": ["strictCamelCase"],
        "selector": "classProperty"
      },
      {
        "format": ["StrictPascalCase"],
        "selector": "classMethod",
        "modifiers": ["public"]
      },
      {
        "format": ["strictCamelCase"],
        "selector": "classMethod"
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expect class members with private identifiers (those starting with `#`) to be considered `private`, so there should be no errors in the example above.

### Actual Result

There are errors in line 3 and 4:
`Class Property name 'privateField' must match one of the following formats: StrictPascalCase`
`Class Method name 'privateMethod' must match one of the following formats: StrictPascalCase`

These errors are due to the rule ignoring the `#` sign and considering the fields as public because there are no typescript accessibility modifiers in front. Using `"filter": { "regex": "^#", "match": false }` to work around the issue also does not work because the `#` sign is ignored in the rule.
